src/Compilers/VisualBasic/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.vb

```diff
@@ -0,0 +1,303 @@
+        End Function
+
+        Public Overrides Function InstrumentFieldOrPropertyInitializer(original As BoundFieldOrPropertyInitializer, rewritten As BoundStatement, symbolIndex As Integer, createTemporary As Boolean) As BoundStatement
+            rewritten = MyBase.InstrumentFieldOrPropertyInitializer(original, rewritten, symbolIndex, createTemporary)
+            Dim syntax As VisualBasicSyntaxNode = original.Syntax
+
+            Select Case syntax.Parent.Parent.Kind()
+                Case SyntaxKind.VariableDeclarator, SyntaxKind.PropertyStatement
+                    Return AddDynamicAnalysis(original, rewritten)
+
+                Case Else
+                    Throw ExceptionUtilities.UnexpectedValue(syntax.Parent.Parent.Kind())
+            End Select
+        End Function
+
+        Public Overrides Function InstrumentForEachLoopInitialization(original As BoundForEachStatement, initialization As BoundStatement) As BoundStatement
+            Return AddDynamicAnalysis(original, MyBase.InstrumentForEachLoopInitialization(original, initialization))
+        End Function
+
+        Public Overrides Function InstrumentForLoopInitialization(original As BoundForToStatement, initialization As BoundStatement) As BoundStatement
+            Return AddDynamicAnalysis(original, MyBase.InstrumentForLoopInitialization(original, initialization))
+        End Function
+
+        Public Overrides Function InstrumentLocalInitialization(original As BoundLocalDeclaration, rewritten As BoundStatement) As BoundStatement
+            Return AddDynamicAnalysis(original, MyBase.InstrumentLocalInitialization(original, rewritten))
+        End Function
+
+        Public Overrides Function CreateUsingStatementPrologue(original As BoundUsingStatement) As BoundStatement
+            Return AddDynamicAnalysis(original, MyBase.CreateUsingStatementPrologue(original))
+        End Function
+
+        Public Overrides Function CreateWithStatementPrologue(original As BoundWithStatement) As BoundStatement
+            Return AddDynamicAnalysis(original, MyBase.CreateWithStatementPrologue(original))
+        End Function
+
+        Private Function AddDynamicAnalysis(original As BoundStatement, rewritten As BoundStatement) As BoundStatement
+            If Not original.WasCompilerGenerated Then
+                Return CollectDynamicAnalysis(original, rewritten)
+            End If
+
+            Return rewritten
+        End Function
+
+        Private Function CollectDynamicAnalysis(original As BoundStatement, rewritten As BoundStatement) As BoundStatement
+            Dim statementFactory As New SyntheticBoundNodeFactory(_factory.TopLevelMethod, _method, original.Syntax, _factory.CompilationState, _diagnostics)
+            Dim analysisPoint As BoundStatement = AddAnalysisPoint(SyntaxForSpan(original), statementFactory)
+            Return If(rewritten IsNot Nothing, statementFactory.StatementList(analysisPoint, rewritten), analysisPoint)
```

Comment:
There are a few constructs where the rewritten statement is optional and so can be Nothing.